src/NHibernate/Multi/QueryBatchItemBase.cs

```diff
@@ -0,0 +1,231 @@
+
+					if (resultsFromCache != null)
+					{
+						// Cached results available, skip the command for them and stores them.
+						qi.Cache = null;
+						_loaderResults[index] = resultsFromCache;
+						continue;
+					}
+				}
+
+				yield return qi.Loader.CreateSqlCommand(qi.Parameters, Session);
+			}
+		}
+
+		public IEnumerable<Func<DbDataReader, int>> GetResultSetHandler()
+		{
+			var dialect = Session.Factory.Dialect;
+			List<object>[] hydratedObjects = new List<object>[_queryInfos.Count];
+
+			for (var i = 0; i < _queryInfos.Count; i++)
+			{
+				Loader.Loader loader = _queryInfos[i].Loader;
+				var queryParameters = _queryInfos[i].Parameters;
+
+				//Skip processing for items already loaded from cache
+				if (_queryInfos[i].CacheKey?.ResultTransformer != null && _loaderResults[i] != null)
+				{
+					_loaderResults[i] = loader.TransformCacheableResults(queryParameters, _queryInfos[i].CacheKey.ResultTransformer, _loaderResults[i]);
+					continue;
+				}
+
+				int entitySpan = loader.EntityPersisters.Length;
+				hydratedObjects[i] = entitySpan == 0 ? null : new List<object>(entitySpan);
+				EntityKey[] keys = new EntityKey[entitySpan];
+
+				RowSelection selection = queryParameters.RowSelection;
+				bool createSubselects = loader.IsSubselectLoadingEnabled;
+
+				_subselectResultKeys[i] = createSubselects ? new List<EntityKey[]>() : null;
+				int maxRows = Loader.Loader.HasMaxRows(selection) ? selection.MaxRows : int.MaxValue;
+				bool advanceSelection = !dialect.SupportsLimitOffset || !loader.UseLimit(selection, dialect);
+
+				var index = i;
+				yield return reader =>
+				{
+					if (advanceSelection)
+					{
+						Loader.Loader.Advance(reader, selection);
+					}
+					if (queryParameters.HasAutoDiscoverScalarTypes)
+					{
+						loader.AutoDiscoverTypes(reader, queryParameters, null);
+					}
+
+					LockMode[] lockModeArray = loader.GetLockModes(queryParameters.LockModes);
+					EntityKey optionalObjectKey = Loader.Loader.GetOptionalObjectKey(queryParameters, Session);
+					int rowCount = 0;
+					var tmpResults = new List<object>();
+
+					int count;
+					for (count = 0; count < maxRows && reader.Read(); count++)
+					{
+						rowCount++;
+
+						object o =
+							loader.GetRowFromResultSet(
+								reader,
+								Session,
+								queryParameters,
+								lockModeArray,
+								optionalObjectKey,
+								hydratedObjects[index],
+								keys,
+								true,
+								_queryInfos[index].CacheKey?.ResultTransformer
+							);
+						if (loader.IsSubselectLoadingEnabled)
+						{
+							_subselectResultKeys[index].Add(keys);
+							keys = new EntityKey[entitySpan]; //can't reuse in this case
+						}
+
+						tmpResults.Add(o);
+					}
+					_loaderResults[index] = tmpResults;
+
+					if (index == _queryInfos.Count - 1)
+					{
+						InitializeEntitiesAndCollections(reader, hydratedObjects);
```

Comment:
It's still can stay here - just be placed outside `yield return' and handle early exits. As I already said I didn't want to leak entities loading related methods to base batch interfaces (and now if I understand correctly it's required anyway for "batchable cache")

Reply:
We do not have the `reader` outside of the yield. The `reader` is a yielded lambda parameter. Storing it in some local variable belonging to the outer scope would be quite a code smell.

The simplest fix seems to me switching to a regular method doing the processing directly in the query item, instead of yielding a callback for letting the batcher finish the work.

Reply:
See in #1789 how I have fixed it. No additional method. Instead, I have change the one yielding those handlers for a method directly doing the job.